tests/integ/test_multithreading.py

```diff
@@ -122,3 +125,67 @@ def test_action_ids_are_unique(session):
             action_ids.add(future.result())
 
     assert len(action_ids) == 10
+
+
+@pytest.mark.parametrize("use_stream", [True, False])
+def test_file_io(session, resources_path, temp_stage, use_stream):
+    stage_prefix = f"prefix_{Utils.random_alphanumeric_str(10)}"
+    stage_with_prefix = f"@{temp_stage}/{stage_prefix}/"
+    test_files = TestFiles(resources_path)
+
+    resources_files = [
+        test_files.test_file_csv,
+        test_files.test_file2_csv,
+        test_files.test_file_json,
+        test_files.test_file_csv_header,
+        test_files.test_file_csv_colon,
+        test_files.test_file_csv_quotes,
+        test_files.test_file_csv_special_format,
+        test_files.test_file_json_special_format,
+        test_files.test_file_csv_quotes_special,
+        test_files.test_concat_file1_csv,
+        test_files.test_concat_file2_csv,
+    ]
+
+    def get_file_hash(fd):
+        return hashlib.md5(fd.read()).hexdigest()
+
+    def put_and_get_file(upload_file_path, download_dir):
+        if use_stream:
+            with open(upload_file_path, "rb") as fd:
+                results = session.file.put_stream(
+                    fd, stage_with_prefix, auto_compress=False, overwrite=False
+                )
+        else:
+            results = session.file.put(
+                upload_file_path,
+                stage_with_prefix,
+                auto_compress=False,
+                overwrite=False,
+            )
+        # assert file is uploaded successfully
+        assert len(results) == 1
+        assert results[0].status == "UPLOADED"
+
+        stage_file_name = f"{stage_with_prefix}{os.path.basename(upload_file_path)}"
+        if use_stream:
+            fd = session.file.get_stream(stage_file_name, download_dir)
+            with open(upload_file_path, "rb") as upload_fd:
+                assert get_file_hash(upload_fd) == get_file_hash(fd)
+
+        else:
+            results = session.file.get(stage_file_name, download_dir)
+            # assert file is downloaded successfully
+            assert len(results) == 1
+            assert results[0].status == "DOWNLOADED"
+            download_file_path = results[0].file
+            # assert two files are identical
+            with open(upload_file_path, "rb") as upload_fd, open(
+                download_file_path, "rb"
+            ) as download_fd:
+                assert get_file_hash(upload_fd) == get_file_hash(download_fd)
+
+    with tempfile.TemporaryDirectory() as download_dir:
+        with ThreadPoolExecutor(max_workers=10) as executor:
+            for file_path in resources_files:
+                executor.submit(put_and_get_file, file_path, download_dir)
```

Comment:
shall we verify all the files are downloaded in the end of the test?

Reply:
we do that check. see lines 178 - 186. Do you want me to add some more checks outside of this?

Reply:
emmm that one is verifying a single file right?
yeah, I'm thinking outside the loop
checking all files show up in the download dir like

os.list(downdir_dir) == [
        test_files.test_file_csv,
        test_files.test_file2_csv,
        test_files.test_file_json,
        test_files.test_file_csv_header,
        test_files.test_file_csv_colon,
        test_files.test_file_csv_quotes,
        test_files.test_file_csv_special_format,
        test_files.test_file_json_special_format,
        test_files.test_file_csv_quotes_special,
        test_files.test_concat_file1_csv,
        test_files.test_concat_file2_csv
]

Reply:
sure. I can add that check as well.

Reply:
added

Reply:
thanks!